modules/admin_manual/pages/configuration/server/security/hsmdaemon.adoc

```diff
@@ -0,0 +1,592 @@
+
+The hsmdaemon is a daemon to delegate encryption to an {hsm-url}[HSM (Hardware Security Module)].
+It can be necessary, as PHP cannot directly interface with {pkcs11-url}[a PKCS11 stack]; neither by an API wrapper, because one does not exist, yet, nor via the OpenSSL bindings
+Because of this, a separate process is needed to decrypt anything with the private key stored in an HSM. 
+
+Running {php-exec-function-url}[exec()] to decrypt the key with a command line command to do the encryption might leak the HSM credentials if the admin lists the currently running processes. 
+To prevent that an "HSM" daemon will be used that can open a session to the HSM upon startup. 
+
+This daemon will be used by ownCloud to decrypt the current master key upon request. 
+The communication happens via {unix-sockets-url}[UNIX sockets] or {network-sockets-url}[TCP sockets] and is authorized by a shared token that the daemon stores in the ownCloud database via a REST/JSON route.
+
+ownCloud internally uses OpenSSL to en-/decrypt keys and needs to be extended to support en-/decrypt operations via the new daemon. 
+The current solution encrypts the ownCloud master key with a key from the HSM. 
+
+NOTE: From the technical point of view the `Crypt` class is extended to handle the key generation in the HSM device and also to get the key from HSM. 
+For the read/write operation on a file, the request goes to the HSM and then, based on the keys fetched from HSM, the files are encrypted or decrypted. 
+The keys are not replaced.
+
+== How The HSM Daemon Interacts with ownCloud
+
+Upon startup, the daemon will generate a token and send it to ownCloud via a new REST/JSON route. 
+After connecting with the HSM daemon, an unsophisticated, line-based, protocol is used (every line ends with CRLF):
+
+. ownCloud sends the token read from database.
+. The daemon compares the received token with its token and returns an `"OK"` line.
+. ownCloud then sends the data it wants to decrypt as a {base64-encoding-url}[Base64-encoded], one-line string.
+. The daemon returns the decrypted data as a Base64-encoded one-line string.
+
+Doing so ensures that an evil admin will need to wiretap the communication between either the database or the HSM daemon and ownCloud.
+
+== Quick Overview
+
+HSM support consists of two core parts:
+
+. An actual HSM PKCS11 module.
+. An hsmdaemon that provides a {jwt-url}[JWT]-protected web API for the PKCS11 stack to generate key pairs and decrypt data.
+
+== Deployment Recommendation
+
+We recommend running hsmdaemon on every web server to reduce latency. 
+
+== Installation
+
+Integrating the hsmdaemon with ownCloud requires 4 steps; these are:
+
+. xref:install-a-pkcs11-module[Install a PKCS11 Module]
+. xref:install-and-configure-the-hsmdaemon[Install and Configure the hsmdaemon]
+. xref:configure-owncloud[Configure ownCloud]
+. xref:install-the-hsmdaemon-as-a-system-service[Install the hsmdaemon as a System Service]
```

Comment:
Question: why is `configure-owncloud` before `install-the-hsmdaemon-as-a-system-service` ?
Shouldnt it be install pkcs11, install hsmdeamon, hsm as system service, configure owncloud?
It may has a reason but the sort order sounds strange

Reply:
Good point. I'm not sure why.